core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnInsertExec.scala

```diff
@@ -78,7 +82,134 @@ case class ColumnInsertExec(_child: SparkPlan, partitionColumns: Seq[String],
+    }
+    defaultBatchSizeTerm = ctx.freshName("defaultBatchSize")
+    ctx.addMutableState("int", defaultBatchSizeTerm, "")
+    val defaultRowSize = ctx.freshName("defaultRowSize")
+    val childProduce = doChildProduce(ctx)
+
+    child match {
+      case c: CallbackColumnInsert =>
+        ctx.addNewFunction(c.resetInsertions,
+          s"""
+             |public final void ${c.resetInsertions}() {
+             |  $batchSizeTerm = 0;
+             |  $numInsertions = -1;
+             |}
+          """.stripMargin)
+        batchBucketIdTerm = Some(c.bucketIdTerm)
+      case _ =>
+    }
+
+    val initEncoderCode =
+      s"""
+         |this.$encoderArrayTerm[i] = $encodingClass$$.MODULE$$.getColumnEncoder(
+         |           |  $schemaTerm.fields()[i]);
+       """.stripMargin
+
+    val initEncoderArray = loop(initEncoderCode, relationSchema.length)
+    val iniCursorCode = s"this.$cursorArrayTerm[i] = 0L;"
+    val iniCursorArray = loop(iniCursorCode, relationSchema.length)
+
+    ctx.addMutableState("org.apache.spark.sql.execution.columnar.encoding.ColumnEncoder[]",
+      encoderArrayTerm,
+      s"""
+         |this.$encoderArrayTerm =
+         | new org.apache.spark.sql.execution.columnar.encoding.ColumnEncoder[$schemaLength];
+         |$initEncoderArray
+        """.stripMargin)
+
+    ctx.addMutableState("long[]", cursorArrayTerm,
+      s"""
+         |this.$cursorArrayTerm = new long[$schemaLength];
+         |$iniCursorArray
+        """.stripMargin)
+
+    val encoderLoopCode = s"$defaultRowSize += " +
+      s"$encoderArrayTerm[i].defaultSize($schemaTerm.fields()[i].dataType());"
+
+    val declarations = loop(encoderLoopCode, relationSchema.length)
+
+    val checkEnd = if (useMemberVariables) {
+      "if (!currentRows.isEmpty()) return"
+    } else {
+      s"if ($numInsertions >= 0) return"
+    }
+    // no need to stop in iteration at any point
+    ctx.addNewFunction("shouldStop",
+      s"""
+         |@Override
+         |protected final boolean shouldStop() {
+         |  return false;
+         |}
+      """.stripMargin)
+    s"""
+       |$checkEnd; // already done
+       |$batchSizeDeclaration
+       |if ($numInsertions < 0) {
+       |  $numInsertions = 0;
+       |  int $defaultRowSize = 0;
+       |  $declarations
+       |  $defaultBatchSizeTerm = Math.max(
+       |    (${math.abs(columnBatchSize)} - 8) / $defaultRowSize, 16);
+       |  // ceil to nearest multiple of $checkFrequency since size is checked
+       |  // every $checkFrequency rows
+       |  $defaultBatchSizeTerm = ((($defaultBatchSizeTerm - 1) / $checkFrequency) + 1)
+       |      * $checkFrequency;
+       |  $initEncoders
+       |  $childProduce
+       |}
+       |if ($batchSizeTerm > 0) {
+       |  $storeColumnBatch(-1, $storeColumnBatchArgs);
+       |  $batchSizeTerm = 0;
+       |}
+       |${
+      if (numInsertedRowsMetric eq null) ""
+      else s"$numInsertedRowsMetric.${metricAdd(numInsertions)};"
+    }
```

Comment:
${} formatting has broken in the 4 lines above. You will need to correct manually since IDEA auto-formatter has trouble with this (and probably introduced these incorrect breaks).